File: export/api/payment_entry.py
import frappe
from frappe.utils import flt


def validate(doc, method=None):
	_apply_exchange_gain_loss_split(doc)


def _apply_exchange_gain_loss_split(doc):
	"""
	Core ERPNext always forces the row flagged `is_exchange_gain_loss` to the
	FULL Paid/Received difference on every save (payment_entry.py:
	set_exchange_gain_loss), which makes it impossible to split that
	difference across more than one account directly in this table.

	This lets the user add extra deduction rows checked as "Split of
	Exchange Gain/Loss" (custom_split_of_exchange_gain_loss). Whatever they
	put in those rows is carved out of the auto-calculated exchange
	gain/loss row, so the total across all rows still matches the actual
	Paid/Received difference exactly - it's just distributed across
	whichever accounts the user chose instead of all going to one.

	Runs after core's own validate() (via the "validate" doc_event), so it
	has the final say on the deduction amounts before save/submit.
	"""
	exchange_gain_loss_row = next((d for d in doc.get("deductions") if d.is_exchange_gain_loss), None)
	if not exchange_gain_loss_row:
		return

	split_rows = [
		d
		for d in doc.get("deductions")
		if d.custom_split_of_exchange_gain_loss and not d.is_exchange_gain_loss
	]
	if not split_rows:
		return

	split_total = sum(flt(d.amount) for d in split_rows)
	if not split_total:
		return

	precision = doc.precision("amount", "deductions")
	remaining = flt(exchange_gain_loss_row.amount) - split_total

	if flt(remaining, precision) < 0:
		frappe.throw(
			frappe._(
				"The amount split across rows marked {label} ({split_total}) is more than "
				"the total Exchange Gain/Loss ({total}). Reduce the split amount so it does "
				"not exceed the total."
			).format(
				label=frappe.bold("Split of Exchange Gain/Loss"),
				split_total=frappe.format_value(split_total, {"fieldtype": "Currency"}),
				total=frappe.format_value(exchange_gain_loss_row.amount, {"fieldtype": "Currency"}),
			),
			title=frappe._("Split Exceeds Total"),
		)

	exchange_gain_loss_row.amount = flt(remaining, precision)

	# core's set_unallocated_amount() treats any deduction row NOT flagged
	# is_exchange_gain_loss as a genuinely separate deduction, which inflates
	# unallocated_amount (and therefore difference_amount) for our split
	# rows too - they're not extra money, just a reclassification of the
	# same exchange gain/loss. Borrow core's own flag briefly so its own
	# calculation excludes them correctly, then put it back: leaving it set
	# would make set_exchange_gain_loss() delete these rows as "extra"
	# exchange-gain-loss rows on the next save.
	for row in split_rows:
		row.is_exchange_gain_loss = 1
	doc.set_unallocated_amount()
	doc.set_difference_amount()
	for row in split_rows:
		row.is_exchange_gain_loss = 0


def round_payment_entry_gl_amounts(doc, method=None):
	"""
	ERPNext computes Payment Entry GL amounts via USD * exchange_rate multiplication
	which produces floating-point residues (e.g. 16992.499986817 instead of 16992.50).
	Round all GL rows for this Payment Entry to 2 decimal places after core on_submit
	has already written them.
	"""
	frappe.db.sql("""
		UPDATE `tabGL Entry`
		SET
			debit                      = ROUND(debit, 2),
			credit                     = ROUND(credit, 2),
			debit_in_account_currency  = ROUND(debit_in_account_currency, 2),
			credit_in_account_currency = ROUND(credit_in_account_currency, 2)
		WHERE voucher_type = 'Payment Entry'
		AND voucher_no = %s
		AND is_cancelled = 0
	""", doc.name)

File: export/public/js/payment_entry.js
/************************************
 * PAYMENT ENTRY — SPLIT EXCHANGE GAIN/LOSS ACROSS ACCOUNTS
 *
 * Core ERPNext always forces the row flagged "is_exchange_gain_loss" to the
 * FULL Paid/Received difference on every recalculation. This mirrors the
 * server-side adjustment (export.api.payment_entry._apply_exchange_gain_loss_split)
 * on the client side too, so the UI doesn't show a stale "full amount" that
 * only corrects itself after save.
 *
 * Usage: on a row with a difference already showing in the auto-created
 * Exchange Gain/Loss row, add another deduction row with the desired
 * account, enter the amount you want carved out, and check "Split of
 * Exchange Gain/Loss" on that row. The Exchange Gain/Loss row's amount
 * will automatically shrink by that much.
 ************************************/
frappe.ui.form.on("Payment Entry", {
    refresh(frm) {
        apply_exchange_gain_loss_split(frm);
    },

    base_paid_amount(frm) {
        setTimeout(() => apply_exchange_gain_loss_split(frm), 300);
    },

    base_received_amount(frm) {
        setTimeout(() => apply_exchange_gain_loss_split(frm), 300);
    },
});

frappe.ui.form.on("Payment Entry Deduction", {
    amount(frm, cdt, cdn) {
        const row = locals[cdt][cdn];
        if (row.is_exchange_gain_loss) return; // avoid reacting to our own adjustment below
        apply_exchange_gain_loss_split(frm);
    },

    custom_split_of_exchange_gain_loss(frm, cdt, cdn) {
        apply_exchange_gain_loss_split(frm);
    },

    deductions_remove(frm) {
        apply_exchange_gain_loss_split(frm);
    },
});

function apply_exchange_gain_loss_split(frm) {
    const rows = frm.doc.deductions || [];
    const exchange_row = rows.find((d) => d.is_exchange_gain_loss);
    if (!exchange_row) return;

    const split_rows = rows.filter((d) => d.custom_split_of_exchange_gain_loss && !d.is_exchange_gain_loss);
    if (!split_rows.length) return;

    const split_total = split_rows.reduce((sum, d) => sum + flt(d.amount), 0);
    if (!split_total) return;

    // Core's own logic resets this row to the FULL paid/received difference on
    // every recalculation, so re-derive from the raw difference each time,
    // not from whatever the row currently shows.
    const full_difference = flt(frm.doc.base_paid_amount) - flt(frm.doc.base_received_amount);
    const remaining = flt(full_difference - split_total, precision("amount", exchange_row));

    if (remaining < 0) {
        frappe.show_alert({
            message: __("Split amount exceeds the total Exchange Gain/Loss — it will be corrected on save."),
            indicator: "red",
        });
        return;
    }

    if (flt(exchange_row.amount) === remaining) return; // already correct, avoid a needless refresh

    frappe.model.set_value(exchange_row.doctype, exchange_row.name, "amount", remaining);
    frm.trigger("set_difference_amount");
}
